arch/risc-v/src/litex/litex_sdio.c

```diff
@@ -0,0 +1,1450 @@
+  DEBUGASSERT(((uint32_t)buffer & 3) == 0);
+
+  mcinfo("buffer: %p bytes: %u\n", buffer, nbytes);
+
+  litex_configxfrints(priv, LITEX_INT_BLOCK2MEM);
+
+  /* flush CPU d-cache */
+
+  up_invalidate_dcache_all();
+
+  putreg32(0, LITEX_SDBLOCK2MEM_DMA_ENABLE);
+  putreg32((uint64_t)(uintptr_t)buffer >> 32, LITEX_SDBLOCK2MEM_DMA_BASE);
+  putreg32((uint64_t)(uintptr_t)buffer, LITEX_SDBLOCK2MEM_DMA_BASE + 0x04);
+  putreg32(nbytes, LITEX_SDBLOCK2MEM_DMA_LENGTH);
+  putreg32(1, LITEX_SDBLOCK2MEM_DMA_ENABLE);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: litex_sendsetup
+ *
+ * Description:
+ *   Setup to perform a write DMA.  If the processor supports a data cache,
+ *   then this method will also make sure that the contents of the DMA memory
+ *   and the data cache are coherent.  For write transfers, this may mean
+ *   flushing the data cache.
+ *
+ * Input Parameters:
+ *   dev    - An instance of the SDIO device interface
+ *   buffer - The memory to DMA into
+ *   nbytes - The size of the DMA transfer in bytes
+ *
+ * Returned Value:
+ *   OK on success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+static int litex_sendsetup(FAR struct sdio_dev_s *dev,
+                              FAR const uint8_t *buffer, size_t nbytes)
+{
+  struct litex_dev_s *priv = (struct litex_dev_s *)dev;
+  DEBUGASSERT(priv != NULL && buffer != NULL && nbytes > 0);
+  DEBUGASSERT(((uint32_t)buffer & 3) == 0);
+
+  mcinfo("buffer: %p bytes: %u\n", buffer, nbytes);
+
+  litex_configxfrints(priv, LITEX_INT_MEM2BLOCK);
+
+  /* flush CPU d-cache */
+
+  up_invalidate_dcache_all();
+
+  putreg32(0, LITEX_SDMEM2BLOCK_DMA_ENABLE);
+  putreg32((uint64_t)(uintptr_t)buffer >> 32, LITEX_SDMEM2BLOCK_DMA_BASE);
+  putreg32((uint64_t)(uintptr_t)buffer, LITEX_SDMEM2BLOCK_DMA_BASE + 0x04);
+  putreg32(nbytes, LITEX_SDMEM2BLOCK_DMA_LENGTH);
+  putreg32(1, LITEX_SDMEM2BLOCK_DMA_ENABLE);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: litex_waitresponse
+ *
+ * Description:
+ *   Poll-wait for the response to the last command to be ready.
+ *
+ * Input Parameters:
+ *   dev  - An instance of the SDIO device interface
+ *   cmd  - The command that was sent.  See 32-bit command definitions above.
+ *
+ * Returned Value:
+ *   OK is success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+static int litex_waitresponse(FAR struct sdio_dev_s *dev, uint32_t cmd)
+{
+  uint32_t ev;
+
+  for (; ; )
+  {
+    if ((cmd & MMCSD_DATAXFR_MASK) == 0)
+      ev = getreg32(LITEX_SDCORE_CMD_EVENT);
+    else
+      ev = getreg32(LITEX_SDCORE_DATA_EVENT);
+
+    mcinfo("%lu\n", ev);
+
+    if (ev & 0x01)
+      break;
+
+    nxsig_usleep(10);
+  }
+
+  /* SD_WRERROR */
+
+  if (ev & 0x02)
```

Comment:
are there any defines for those numbers?

Reply:
Yes there is. Added.